Return Error for invalid identifiers instead of Warning

ceco556 opened the pull request:
## Changes proposed by this PR

closes #9271
closes https://github.com/concourse/concourse/issues/8725
closes https://github.com/concourse/concourse/issues/8475

* [x] When there is invalid identifiers in a fly command it will return Error instead of Warning.
* [x] Check for invalid identifiers in the set_pipeline step and return Error if there is. 

## Notes to reviewer

I changed Warnings to be ConfigErrors. Let me know if you are okay with this. If needed i will rename it to IdentifierErrors or something else.

## Release Note

Fly commands now will return Error when there is invalid identifiers found.
If there is set_pipeline step with invalid identifier in a pipeline and you try to set it, it will return Error.

analytically committed:
atc: enhance container memory unit parsing with IEC notation support

Add support for IEC standard notation (KiB, MiB, GiB) alongside the existing
traditional units (KB, MB, GB). Improve the regex pattern to handle abbreviated
units (K, M, G) and ensure case insensitivity across all unit formats. No
functional change to existing behavior - all units continue to use binary
multipliers (2^10, 2^20, 2^30) as before.

Signed-off-by: Mathias Bogaert <mathias.bogaert@gmail.com>

analytically committed:
fix: add missing O_TRUNC flag to file write operations

When writing metadata files, the code was using O_WRONLY|O_CREATE flags without
O_TRUNC. This could cause file corruption when writing shorter content than what
existed previously, as the remaining bytes from the old content would persist in
the file.

The fix adds O_TRUNC to ensure clean overwrites and maintain data integrity.

Signed-off-by: Mathias Bogaert <mathias.bogaert@gmail.com>

IvanChalukov committed:
Implement migration from md5 hashing to sha256

Signed-off-by: IvanChalukov <ichalukov@gmail.com>

IvanChalukov committed:
clean up clock_timestamp

Signed-off-by: IvanChalukov <ichalukov@gmail.com>

IvanChalukov committed:
Update resource_cache_collector_test to use sha256 hashing

Signed-off-by: IvanChalukov <ichalukov@gmail.com>

Kump3r committed:
Adding the db connection parameter

Signed-off-by: Kump3r <tonevkalin@gmail.com>

taylorsilva committed:
update links in README

Signed-off-by: Taylor Silva <dev@taydev.net>

ceco556 committed:
Return error for invalid identifier

ceco556 committed:
Fix tests

Fix

Fix

New changes

Fix

Fixes

Fix tests

Signed-off-by: ceco556 <ceco456@gmail.com>

ceco556 commented:
I have only on question here. Do we need a fallback mechanism for the delete-pipeline, delete-team and ge-pipeline?
If there are already set pipelines with invalid identifiers and try to delete them with this change it will return Error.

For me it would be good to have such for some time. @taylorsilva , what do you think?

ceco556 commented:
@taylorsilva , could you please give me access to the PR checks so i can fix them?

ceco556 committed:
Merge branch 'concourse:master' into master

Signed-off-by: ceco556 <ceco456@gmail.com>

ceco556 commented:
Sorry for the commits, my forked repo was not synced.
If it is a problem i will open new PR.

taylorsilva commented:
Initial thought for the users that have pipelines with invalid identifiers, since I think the validation only happens client side, would be to add a flag to skip the validation check, `--skip-validation`.

This would be a final escape hatch for users that have ignored the warnings.

ceco556 commented:
I will implement it. Thanks for the suggestion.

After log out and log in, still i don't see the PR checks. Other pipelines are visible now. Did i miss something?

ceco556 commented:
Ah, yes, found it!
Thank you!

taylorsilva commented:
Ignore the `[containerd-integration](https://ci.concourse-ci.org/builds/464632242)` failure. It's flaking, I'll get it passing.

The unit test failures are legit though: https://ci.concourse-ci.org/builds/464630948#L6821cb41:627:654

Commit:
Fix unit tests

Commit:
Fix rename_pipeline tests

ceco556 commented:
The unit job timed out right at the end. I've retriggered it. Is it normal to time out ?

I think i managed to fix the unit test, now i have to take a look at the testflight tests.

taylorsilva commented:
Sadly the time out is normal. There's only one worker for all PR's right now which creates a bottleneck and these timeouts are one side effect. Hopefully we can get more PR workers soon.

ceco556 commented:
While fixing the testflight tests i found one issue. If you try to set a pipeline with set_pipeline step in it such as:

```
jobs:
- name: sp-1
  public: true
  plan:
    - get: some-resource
    - set_pipeline: ((pipeline_name))
      file: some-resource/pipeline.yml
      var_files:
      - some-resource/name.yml
      vars:
        greetings: hello
```
It will return error because the name ((pipeline_name)) has invalid identifier - "(". 
My suggestion here is to change the regex to not check for "(". What do you think?

Commit:
Fix validation for var placeholders and fix testflight tests

taylorsilva commented on atc/step_validator.go:
```diff
@@ -81,19 +74,12 @@ func (validator *StepValidator) VisitTask(plan *TaskStep) error {
 		})
 	}
 
-	if plan.Hermetic {
-		validator.recordWarning(ConfigWarning{
-			Type:    "pipeline",
-			Message: validator.annotate("specifies `hermetic:` only works against worker containerd runtime"),
-		})
-	}
```
oh this is unfortunate, but I don't think we should keep the warning stuff just for this one warning. We warn about this in the docs already.
I'm mostly commenting so folks know I saw this and accepted that we won't issue this warning for tasks that set `hermetic` anymore.

taylorsilva commented on atc/step_validator.go:
```diff
@@ -223,12 +196,9 @@ func (validator *StepValidator) VisitSetPipeline(step *SetPipelineStep) error {
-	}
-	if warning != nil {
-		validator.recordWarning(*warning)
+	configError := ValidateIdentifier(step.Name, validator.context...)
+	if configError != nil {
+		validator.recordConfigErrors(*configError)
```
I'm guessing it's here that we're getting the error when the step name is a var:
```yaml
- set_pipeline: ((pipeline_name))
  file: pipeline.yml
```

Let me know what you think of this. Before calling `ValidateIdentifier()` on the `step.Name`, check if it contains any vars in the name (you could re-use your regex here!). In that case we don't call `ValidateIdentifier()` and leave the pipeline name to be checked later during step runtime after the var(s) are resolved.

If the `step.Name` does not contain any vars, then we proceed with calling `ValidateIdentifier()` on it and returning any errors.

This allows us to provide users with early feedback about the pipeline name when they're not using vars, and still allowing users to use vars in the pipeline name if they choose to.

Wdyt?

ceco556 replied:
I think this is a good idea. 
Will fix it.

taylorsilva commented on atc/step_validator.go:
I think we an remove `ConfigErrors` and just use `Errors`. We could then remove the `recordConfigErrors()` function and only use `recordError()` everywhere.

taylorsilva commented on fly/commands/internal/setpipelinehelpers/atc_config.go:
```diff
@@ -28,7 +28,7 @@ type ATCConfig struct {
 	CheckCredentials bool
 	DryRun           bool
-	CommandWarnings  []concourse.ConfigWarning
+	CommandErrors    []atc.ConfigErrors
```
I don't think this field is will be needed anymore if you consolidate to just return `error` instead of using `[]atc.ConfigErrors` everywhere still.

taylorsilva commented on fly/commands/set_pipeline.go:
```diff
@@ -34,26 +35,37 @@ type SetPipelineCommand struct {
 
-func (command *SetPipelineCommand) Validate() ([]concourse.ConfigWarning, error) {
-	var warnings []concourse.ConfigWarning
+func (command *SetPipelineCommand) Validate() ([]atc.ConfigErrors, error) {
+	var configErrors []atc.ConfigErrors
```
dup: consolidate to return `error` only

```go
func (command *SetPipelineCommand) Validate() error {
```

taylorsilva commented on fly/integration/error_handling_test.go:
```diff
@@ -101,10 +101,8 @@ var _ = Describe("Fly CLI", func() {
-				sess, err := gexec.Start(flyCmd, GinkgoWriter, GinkgoWriter)
-				Expect(err).NotTo(HaveOccurred())
-
-				Eventually(sess.Err.Contents).Should(ContainSubstring(`error: team 'doesnotexist' does not exist`))
```
Can we keep the `ContainSubstring()` checks in these tests?

taylorsilva commented on fly/integration/rename_pipeline_test.go:
```diff
@@ -164,11 +163,10 @@ var _ = Describe("RenamePipeline", func() {
+			Expect(err).ToNot(HaveOccurred())
 
-			Eventually(sess).Should(gexec.Exit(0))
-			Expect(sess.Out).To(gbytes.Say(fmt.Sprintf("pipeline successfully renamed to '%s'", newName)))
```
nit: can we keep the output check here?

taylorsilva commented on go-concourse/concourse/teams_test.go:
```diff
@@ -174,18 +174,15 @@ var _ = Describe("ATC Handler Teams", func() {
 					_, _, _, warnings, err := team.CreateOrUpdate(desiredTeam)
 					Expect(err).NotTo(HaveOccurred())
-					Expect(warnings).To(HaveLen(1))
-					Expect(warnings[0].Message).To(ContainSubstring("team: 'new venture' is not a valid identifier"))
```
nit: can we still check for the string?

taylorsilva reviewed (changes requested):
Overall we're on the right track here.

The majority of my feedback basically boils down to: delete more code! 😀 

I think anywhere where we were returning an `atc.ConfigWarning` can be consolidated into the `error` that was returned along with the `ConfigWarning`. That should result in code being deleted and simplify method signatures to return only `error` instead of warnings and errors.

Also great job adding the `SkipValidation` escape hatch for various fly commands 👏

ceco556 commented:
Hi @taylorsilva,

Thanks for the code review. While addressing your comments i encountered with the integration tests for fly.(set_pipeline_test.go)
I get a lot of errors for invalid groups, resources, etc. Should we raise errors for these cases?
If yes, could you please assist with the test. Let me know if you'd like to commit my latest changes.

taylorsilva commented:
Hey, FINALLY circling back to this. So it seems our biggest issue is with handling vars `(( ))` in identifiers. You [previously shared](https://github.com/concourse/concourse/pull/9202#issuecomment-2898603607) this example:

```yaml
jobs:
- name: sp-1
  public: true
  plan:
    - get: some-resource
    - set_pipeline: ((pipeline_name))
      file: some-resource/pipeline.yml
      var_files:
      - some-resource/name.yml
      vars:
        greetings: hello
```

What confuses me about this is: when is `((pipeline_name))` expected to be interpolated?
I _think_ `((pipeline_name))` would be interpolated when this pipeline is set. If I make your example a bit more portable:

```yaml
resources:
- name: mock
  type: mock
  source:
    create_files:
      pipeline.yml: |
        resources:
          - name: mock
            type: mock
        jobs:
          - name: job
            plan:
              - get: mock

jobs:
- name: sp
  plan:
    - get: mock
    - set_pipeline: ((pipeline_name))
      file: mock/pipeline.yml
```
and `fly set-pipeline` and run it, I get an error `undefined vars: pipeline_name`
<img width="308" height="121" alt="image" src="https://github.com/user-attachments/assets/fea6cef1-4491-4edd-8ab1-b775840bddf9" />
This error is interesting for other reasons, but the main point for now is that it fails.

If I set this pipeline again with `--var pipeline_name=sp-test-ch` set, everything works:
<img width="244" height="101" alt="image" src="https://github.com/user-attachments/assets/2ed242a3-66a4-411a-883e-7d604a061814" />

What workflow were you conveying when you shared this pipeline?

---

I decided to see where vars currently work through out a pipeline. I think I found some bugs related to how and when Concourse interpolates vars, but I think we can get away with ignoring those for now for the sake of this PR.

First, there are three "types" of vars that Concourse can interpolate:

- `(( ))` - the classic. Can be interpolated when a pipeline is being set and used anywhere in a pipeline. Can also be interpolated during runtime, where the value is then expected to come from the configured secrets manager. Runtime interpolation is limited though (e.g. step names cannot be interpolated, but a `task.config` can be)
- `((.:))` - the "local" var source, created by `load_var` and `across` steps
- `((src:))` - var sources, come from `var_sources` defined in a pipeline

I created the following pipeline to play around with these different types of vars in various situations.

<details>
<summary>Click to view pipeline</summary>

```yaml
var_sources:
  - name: varsrc
    type: dummy
    config:
      vars:
        task-name: my-vars
        child-pipe: child_pipe

resources:
  - name: mock
    type: mock

jobs:
  - name: job
    plan:
      - get: image
        resource: mock
        params:
          mirror_self_via_params: true

      - get: mock
        params:
          create_files_via_params:
            vars.yml: |
              task_name: some_task_name
              invalid_task_name: invalid task name
              child-pipe: child_pipe
            pipe.yml: |
              resources:
                - name: mock
                  type: mock
              jobs:
                - name: job
                  plan:
                    - get: mock

      - load_var: my-vars
        reveal: true
        file: mock/vars.yml

      - across:
          - var: acr
            values:
              - 123
              - abc
        in_parallel:
          # is currently allowed and should remain so
          # should we error if the identifier is not valid???
          - task: ((.:acr))
            image: image
            config:
              platform: linux
              run:
                path: bash
                args:
                  - -ec
                  # referncing var_sources does not work ((varsrc:child-pipe)) inside across steps
                  # results in error: missing source 'varsrc' in var: varsrc:child-pipe
                  # referencing ((.:my-vars)) does work! it is interpolated
                  - |
                    echo ((.:my-vars.child-pipe))-((.:acr))
          # does not warn, will create a pipeline with the literal var `(( ))` as the name
          # ((.:my-vars)) is not interpolatd
          # ((.:acr)) is interpolated
          # ((varsrc:)) results in an error during runtime: missing source 'varsrc'. I think this means lookups from secrets manager may also fail...
          - set_pipeline: ((.:my-vars.child-pipe))-((.:acr))
            file: mock/pipe.yml

      # currently warns, and should not be allowed moving forward
      - task: ((.:task-name))
        image: image
        config:
          platform: linux
          run:
            path: bash
            args:
              - -ec
              - |
                echo ((.:my-vars.child-pipe))-((varsrc:child-pipe))

      # it seems we detect the var, but it's never actually interpolated. We end
      # up with a bad pipeline name. The step should error.
      - set_pipeline: ((.:my-vars.child-pipe))
        file: mock/pipe.yml
```

</details>

There are a number of different cases we have to deal with here. Overall, I found the variable interpolation within `across` steps isn't working as expected. There are some bugs here. Thankfully, the core part works as expected and I think the fact that no one has logged any of the bugs I just found means I've stumbled across a very unusual use-case that we can ignore for now. Read my comments in the pipeline above for details on what works and doesn't work throughout the pipeline.

What does this all mean for this PR? There are various places where "validation" happens and we can check identifiers. I think when setting a pipeline, we can ignore any vars found inside `across` steps. Any vars found in identifiers outside `across` steps should result in an error.

During runtime, I think I would like to see the `across` properly validate identifiers for `task` and `set_pipeline` steps. How both of those are handled may be different. My expertise in that area of the code base is a bit foggy. I consider this part a bit out of scope for this PR, feels a bit like scope-creep potentially.

Those are my thoughts right now. I'm interested in your response to the first section of my post here. Let me know your thoughts! Sorry for the longggg post 😬
